File: ocaml/lib/kan/day.ml
type ('f, 'g) day = forall 'b 'c. 'f 'b * 'g 'c * ('b -> 'c -> 'a)

let day (f : 'a -> 'b) (g : 'c -> 'd) : ('a -> 'c, 'b, 'd) day =
  (f, g, fun a c -> (a, c))

let fmap (f : 'a -> 'b) (day : ('f, 'g) day) : ('f, 'g) day =
  let (fb, gc, bca) = day in
  (fb, gc, fun b c -> f (bca b c))

let pure (x : 'a) : ('f, 'g) day =
  ((), (), fun _ _ -> x)

let ap (day1 : ('f, 'g) day) (day2 : ('f, 'g) day) : ('f, 'g) day =
  let (fa, fb, u) = day1 in
  let (gc, gd, v) = day2 in
  ((fa, gc), (fb, gd), fun (a, c) (b, d) -> u a b (v c d))

let assoc (day : ('f, ('g, 'h) day) day) : (('f, 'g), 'h) day =
  let (fb, (gd, he, dec), bca) = day in
  ((fb, gd), he, fun (b, d) e -> bca b (dec d e))

let disassoc (day : (('f, 'g), 'h) day) : ('f, ('g, 'h) day) =
  let ((fb, gc), hd, eda) = day in
  (fb, (gc, hd, fun c d -> eda (fb, c) d))

let swapped (day : ('f, 'g) day) : ('g, 'f) day =
  let (fb, gc, abc) = day in
  (gc, fb, fun c b -> abc b c)

let intro1 (f : 'a -> 'b) : ('unit, 'a -> 'b) day =
  ((), f, fun _ a -> a)

let intro2 (f : 'a -> 'b) : ('a -> 'b, 'unit) day =
  (f, (), fun a _ -> a)

let elim1 (day : ('unit, 'f) day) : 'f =
  let ((), fc, bca) = day in
  bca () fc

let elim2 (day : ('f, 'unit) day) : 'f =
  let (fb, (), bca) = day in
  fb

let trans1 (fg : 'a -> 'b) (day : ('a, 'g) day) : ('b, 'g) day =
  let (fb, hc, bca) = day in
  (fg fb, hc, bca)

let trans2 (gh : 'a -> 'b) (day : ('f, 'a) day) : ('f, 'b) day =
  let (fb, gc, bca) = day in
  (fb, gh gc, bca)

File: python/kan/day.py
from typing import Callable, Generic, TypeVar

F = TypeVar('F')
G = TypeVar('G')
A = TypeVar('A')
B = TypeVar('B')

class Day(Generic[F, G, A]):
    def __init__(self, f: F, g: G, apply: Callable[[F, G], A]):
        self.f = f
        self.g = g
        self.apply = apply

    def fmap(self, f: Callable[[A], B]) -> 'Day[F, G, B]':
        return Day(self.f, self.g, lambda ff, gg: f(self.apply(ff, gg)))

    @staticmethod
    def pure(x: A) -> 'Day[None, None, A]':
        return Day(None, None, lambda _, __: x)

    def ap(self, other: 'Day[F, G, Callable[[A], B]]') -> 'Day[F, G, B]':
        return Day(
            (self.f, other.f),
            (self.g, other.g),
            lambda ff, gg: other.apply(ff[0], gg[0])(self.apply(ff[1], gg[1]))
        )

def assoc(day: Day[F, Day[G, H, A], A]) -> Day[Day[F, G, A], H, A]:
    return Day(
        Day(day.f, day.g.f, lambda ff, gg: day.apply(ff, (gg, day.g.g))),
        day.g.g,
        lambda fga, hc: day.g.apply(fga.g, hc)(fga.apply(fga.f, fga.g))
    )

def disassoc(day: Day[Day[F, G, A], H, A]) -> Day[F, Day[G, H, A], A]:
    return Day(
        day.f.f,
        Day(day.f.g, day.g, lambda gg, hh: day.apply((gg, hh.f), hh.g)),
        lambda ff, gha: day.apply((ff, gha.f), gha.g)(gha.apply(gha.f, gha.g))
    )

def swapped(day: Day[F, G, A]) -> Day[G, F, A]:
    return Day(day.g, day.f, lambda gg, ff: day.apply(ff, gg))

def intro1(f: F) -> Day[None, F, A]:
    return Day(None, f, lambda _, ff: ff)

def intro2(g: G) -> Day[G, None, A]:
    return Day(g, None, lambda gg, _: gg)

def elim1(day: Day[None, F, A]) -> F:
    return day.g

def elim2(day: Day[G, None, A]) -> G:
    return day.f

def trans1(f: Callable[[A], B], day: Day[F, G, A]) -> Day[F, G, B]:
    return Day(day.f, day.g, lambda ff, gg: f(day.apply(ff, gg)))

def trans2(g: Callable[[G], H], day: Day[F, G, A]) -> Day[F, H, A]:
    return Day(day.f, g(day.g), lambda ff, hh: day.apply(ff, hh))